activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb

```diff
@@ -131,8 +131,8 @@ def exec_query(sql, name = "SQL", binds = [], prepare: false)
       # Executes insert +sql+ statement in the context of this connection using
       # +binds+ as the bind substitutes. +name+ is logged along with
       # the executed +sql+ statement.
-      def exec_insert(sql, name = nil, binds = [], pk = nil, sequence_name = nil)
-        sql, binds = sql_for_insert(sql, pk, binds)
+      def exec_insert(sql, name = nil, binds = [], pk = nil, sequence_name = nil, returning = [])
```

Comment:
given that returning is optional, I suspect the extra allocation here is undesired

as an aside, I believe returning would be relevant to handling composite primary keys

https://github.com/composite-primary-keys/composite_primary_keys/blob/c5c27220e844d95e568728dd5fde705e83adf814/lib/composite_primary_keys/connection_adapters/postgresql/database_statements.rb#L12-L18

Reply:
Hi @bf4 , the reason returning is sent here is for the values to make it's way into the [place where the sql statement is build](https://github.com/rails/rails/pull/39968/files/543627f47cbb69d0007e7f43c94372c18f2d4d4d#diff-e6fd03cad5e3437c76b6c5db106359124dc9feab8341ade39b9ae54af001fac9R86)

Reply:
I meant allocating the array when its use is the exception in what I guess is a hotspot is worth considering the non-nil default

Reply:
Yeah, the suggestion would be to do `returning = nil`, and then later in spots where you expect it to be an array, account for it to be `nil` as well.

This means one less array for queries that don't use the returning feature, and that extra array can add up for a function that's called a lot in a web request.

Reply:
I was following this logic:
- String default: nil (like `pk`)
- Array default: empty array (like `bind`)

But I see what you mean @bf4 @natematykiewicz and will do the change. Thanks for the suggestion.